distributed/tests/test_worker.py

```diff
@@ -3502,3 +3505,43 @@ async def test_reconnect_argument_deprecated(s):
         warnings.simplefilter("error")
         async with Worker(s.address):
             pass
+
+
+@gen_test()
+async def test_worker_makes_own_thread():
+    class InitWorkerNewThread(WorkerPlugin):
+        name: str
+        thread: threading.Thread
+
+        def __init__(self, scheduler):
+            super().__init__()
+            self.scheduler = scheduler
+
+        def setup(self, worker):
+            self.loop = asyncio.new_event_loop()
+            self.thread = threading.Thread(target=self.loop.run_forever)
+            self.thread.start()
+            future = asyncio.run_coroutine_threadsafe(self.func(), self.loop)
+            future.result()
+
+        def teardown(self, worker):
+            self.loop.call_soon_threadsafe(self.loop.stop)
+            self.thread.join(timeout=10)
+
+        async def func(self):
+            try:
+                worker = get_worker()
```

Comment:
I'm not sure what the use-case for this is. `get_worker()` will *always* fail here for freshly started workers because the worker is in "worker.status == "init"`

but because you're in a worker plugin you know the worker will end up running after your setup method finishes

Reply:
if that's needed then the test case should look like:

```python
@gen_cluster(client=True, nthreads=[])
async def test_worker_running_before_running_plugins(c, s, caplog):
    class InitWorkerNewThread(WorkerPlugin):
        name: str = "init_worker_new_thread"
        setup_status: Status | None = None

        def setup(self, worker):
            self.setup_status = worker.status

        def teardown(self, worker):
            pass

    await c.register_worker_plugin(InitWorkerNewThread())
    async with Worker(s.address) as worker:
        await c.submit(inc, 1) == 2
        assert worker.plugins[InitWorkerNewThread.name].setup_status is Status.running```

Reply:
> I'm not sure what the use-case for this is. currently`get_worker()` will _always_ fail here for freshly started workers because the worker is in "worker.status == "init"`
> 
> but because you're in a worker plugin you know the worker will end up running after your setup method finishes

Fair point, the test I made replicated the error I saw, but wasn't exactly the same as we were seeing in our code. I've changed the test now to be more like our code base. 

The use-case is this: the function we are calling isn't part of the worker plugin (so I've moved it out now) but rather initialises an app. This app is initialised once on the host machine, and then once on each worker. We use get_worker to determine if we need to register the worker plugin or not. However, due to recent changes in how worker status is handled, we were not finding workers with get_worker, and so registered the plugin again when it shouldn't be. I assume that get_worker not finding our workers has something to do with that our workers initialise their own threads, but that's just my best guess. Either way, reverting the commented out status change fixes this for us (not saying that's the best solution, just the only one I've found so far and as that line was commented out rather than deleted, I thought it might be fine to put it back...).

Reply:
> if that's needed then the test case should look like:
> 
> ```python
> @gen_cluster(client=True, nthreads=[])
> async def test_worker_running_before_running_plugins(c, s, caplog):
>     class InitWorkerNewThread(WorkerPlugin):
>         name: str = "init_worker_new_thread"
>         setup_status: Status | None = None
> 
>         def setup(self, worker):
>             self.setup_status = worker.status
> 
>         def teardown(self, worker):
>             pass
> 
>     await c.register_worker_plugin(InitWorkerNewThread())
>     async with Worker(s.address) as worker:
>         await c.submit(inc, 1) == 2
>         assert worker.plugins[InitWorkerNewThread.name].setup_status is Status.running
> ```

This makes a lot of sense. Thank you!